### PR TITLE
lithium: skip ticking block entities that are doing nothing 

### PR DESCRIPTION
**Pull request by Titaniumtown**
This PR ports an optimization from lithium that skips ticking block entities that aren't doing anything (as the title says). I've been testing this for a little over 2 weeks now without any issues, more testing is probably needed in regard to plugins.

### CODE REVIEW

**Comment by Titaniumtown** [comment]
Todo: add config to disable this optimization. (in case it causes issues in the future)

File: patches/server/0067-lithium-skip-ticking-block-entities-that-are-doing-n.patch [hunk 1]
```diff
@@ -0,0 +1,1348 @@
...
+
+diff --git a/pom.xml b/pom.xml
+index c48503c42e37545ed606f6a6088ca06d057787ce..c58fde8014775216c57b46011021b1962eb15f38 100644
+--- a/pom.xml
++++ b/pom.xml
+@@ -35,6 +35,12 @@
+      <!-- Yatopia end -->
+ 
+     <dependencies>
++        <!-- https://mvnrepository.com/artifact/org.jetbrains/annotations -->
```
Comment (ishland) [inline patches/server/0067-lithium-skip-ticking-block-entities-that-are-doing-n.patch]: Use built-in javax annotation

File: patches/server/0067-lithium-skip-ticking-block-entities-that-are-doing-n.patch [hunk 2]
```diff
@@ -0,0 +1,1348 @@
...
+\ No newline at end of file
+diff --git a/src/main/java/net/minecraft/server/TileEntity.java b/src/main/java/net/minecraft/server/TileEntity.java
+index c0d465749093b1a18bd5179eb28ae06a00850185..db1fee8735bdecf24070699f7fc4284eabb25faf 100644
+--- a/src/main/java/net/minecraft/server/TileEntity.java
++++ b/src/main/java/net/minecraft/server/TileEntity.java
+@@ -14,8 +14,9 @@ import org.bukkit.inventory.InventoryHolder;
+ import org.spigotmc.CustomTimingsHandler; // Spigot
+ import co.aikar.timings.MinecraftTimings; // Paper
+ import co.aikar.timings.Timing; // Paper
++import me.jellysquid.mods.lithium.common.world.blockentity.SleepingBlockEntity;
```
Comment (ishland) [inline patches/server/0067-lithium-skip-ticking-block-entities-that-are-doing-n.patch]: // Yatopia

File: patches/server/0067-lithium-skip-ticking-block-entities-that-are-doing-n.patch [hunk 3]
```diff
@@ -0,0 +1,1348 @@
...
+     // CraftBukkit start - data containers
+diff --git a/src/main/java/net/minecraft/server/TileEntityBeehive.java b/src/main/java/net/minecraft/server/TileEntityBeehive.java
+index 69ef94e887fa9da1b3ddaf0a345fecc49d2866b5..42cd92f4cfa504612417fdd09c9b894077ba9f71 100644
+--- a/src/main/java/net/minecraft/server/TileEntityBeehive.java
++++ b/src/main/java/net/minecraft/server/TileEntityBeehive.java
+@@ -4,6 +4,9 @@ import com.google.common.collect.Lists;
+ import java.util.Iterator;
+ import java.util.List;
+ import javax.annotation.Nullable;
++import me.jellysquid.mods.lithium.common.util.collections.ListeningList;
```
Comment (ishland) [inline patches/server/0067-lithium-skip-ticking-block-entities-that-are-doing-n.patch]: // Yatopia

File: patches/server/0067-lithium-skip-ticking-block-entities-that-are-doing-n.patch [hunk 4]
```diff
@@ -0,0 +1,1348 @@
...
++import me.jellysquid.mods.lithium.common.world.blockentity.BlockEntitySleepTracker;
```
Comment (ishland) [inline patches/server/0067-lithium-skip-ticking-block-entities-that-are-doing-n.patch]: // Yatopia

File: patches/server/0067-lithium-skip-ticking-block-entities-that-are-doing-n.patch [hunk 5]
```diff
@@ -0,0 +1,1348 @@
...
++import java.util.ArrayList;
```
Comment (ishland) [inline patches/server/0067-lithium-skip-ticking-block-entities-that-are-doing-n.patch]: // Yatopia

File: patches/server/0067-lithium-skip-ticking-block-entities-that-are-doing-n.patch [hunk 6]
```diff
@@ -0,0 +1,1348 @@
...
+ 
+     public TileEntityBell() {
+         super(TileEntityTypes.BELL);
+@@ -20,6 +23,11 @@ public class TileEntityBell extends TileEntity implements ITickable {
+ 
+     @Override
+     public boolean setProperty(int i, int j) {
++       // Yatopia start - lithium: skip ticking block entities that are doing nothing
++       if (!this.ringing && i == 1 && this.world != null) {
++            ((BlockEntitySleepTracker) this.world).setAwake(this, true);
```
Comment (ishland) [inline patches/server/0067-lithium-skip-ticking-block-entities-that-are-doing-n.patch]: Don't use mixin interface workaround

File: patches/server/0067-lithium-skip-ticking-block-entities-that-are-doing-n.patch [hunk 7]
```diff
@@ -0,0 +1,1348 @@
...
+ 
+diff --git a/src/main/java/net/minecraft/server/TileEntityBell.java b/src/main/java/net/minecraft/server/TileEntityBell.java
+index 6963f3c8a0e257615084b558f5ce287aab9722ff..04af18f97921d52252eb89a70f751d2fb604f8a9 100644
+--- a/src/main/java/net/minecraft/server/TileEntityBell.java
++++ b/src/main/java/net/minecraft/server/TileEntityBell.java
+@@ -3,6 +3,7 @@ package net.minecraft.server;
+ import java.util.Iterator;
+ import java.util.List;
+ import org.apache.commons.lang3.mutable.MutableInt;
++import me.jellysquid.mods.lithium.common.world.blockentity.BlockEntitySleepTracker;
```
Comment (ishland) [inline patches/server/0067-lithium-skip-ticking-block-entities-that-are-doing-n.patch]: // Yatopia

File: patches/server/0067-lithium-skip-ticking-block-entities-that-are-doing-n.patch [hunk 8]
```diff
@@ -0,0 +1,1348 @@
...
+diff --git a/src/main/java/net/minecraft/server/TileEntityCampfire.java b/src/main/java/net/minecraft/server/TileEntityCampfire.java
+index 3239d8dd120fda2faa6b770603c5f613c13f7c43..ff301c8c212b38ee9a4b84ad08f393d5947d2eb2 100644
+--- a/src/main/java/net/minecraft/server/TileEntityCampfire.java
++++ b/src/main/java/net/minecraft/server/TileEntityCampfire.java
+@@ -9,12 +9,18 @@ import org.bukkit.craftbukkit.block.CraftBlock;
+ import org.bukkit.craftbukkit.inventory.CraftItemStack;
+ import org.bukkit.event.block.BlockCookEvent;
+ // CraftBukkit end
++import net.minecraft.server.NonNullList;
++import me.jellysquid.mods.lithium.common.world.blockentity.BlockEntitySleepTracker;
```
Comment (ishland) [inline patches/server/0067-lithium-skip-ticking-block-entities-that-are-doing-n.patch]: // Yatopia

File: patches/server/0067-lithium-skip-ticking-block-entities-that-are-doing-n.patch [hunk 9]
```diff
@@ -0,0 +1,1348 @@
...
+@@ -7,14 +7,27 @@ import java.util.List;
+ import org.bukkit.craftbukkit.entity.CraftHumanEntity;
+ import org.bukkit.entity.HumanEntity;
+ // CraftBukkit end
++ import me.jellysquid.mods.lithium.common.world.blockentity.BlockEntitySleepTracker; // Yatopia
+ 
+ public class TileEntityChest extends TileEntityLootable { // Paper - Remove ITickable
+ 
+     private NonNullList<ItemStack> items;
+     protected float a;
```
Comment (ishland) [inline patches/server/0067-lithium-skip-ticking-block-entities-that-are-doing-n.patch]: Is it necessary to do extra logic if chest isnt tickable?

File: patches/server/0067-lithium-skip-ticking-block-entities-that-are-doing-n.patch [hunk 10]
```diff
@@ -0,0 +1,1348 @@
...
+     public boolean setProperty(int i, int j) {
++        this.checkWakeUp(); // Yatopia - lithium: skip ticking block entities that are doing nothing
+         if (i == 1) {
+             this.viewingCount = j;
+             return true;
+@@ -228,7 +258,7 @@ public class TileEntityChest extends TileEntityLootable { // Paper - Remove ITic
+             // CraftBukkit end
+             this.onOpen();
+         }
+-
```
Comment (ishland) [inline patches/server/0067-lithium-skip-ticking-block-entities-that-are-doing-n.patch]: Unnecessary diff

File: patches/server/0067-lithium-skip-ticking-block-entities-that-are-doing-n.patch [hunk 11]
```diff
@@ -0,0 +1,1348 @@
...
+         EntityHuman entityhuman = this.world.a((double) this.position.getX() + 0.5D, (double) this.position.getY() + 0.5D, (double) this.position.getZ() + 0.5D, 3.0D, false);
+diff --git a/src/main/java/net/minecraft/server/TileEntityEnderChest.java b/src/main/java/net/minecraft/server/TileEntityEnderChest.java
+index 80440aa24f546dd151e89b705322d9c3e906b139..94c2bfce5059d784655d729f384c70b165432311 100644
+--- a/src/main/java/net/minecraft/server/TileEntityEnderChest.java
++++ b/src/main/java/net/minecraft/server/TileEntityEnderChest.java
+@@ -1,17 +1,42 @@
+ package net.minecraft.server;
++import me.jellysquid.mods.lithium.common.world.blockentity.BlockEntitySleepTracker;
+ 
+ public class TileEntityEnderChest extends TileEntity { // Paper - Remove ITickable
```
Comment (ishland) [inline patches/server/0067-lithium-skip-ticking-block-entities-that-are-doing-n.patch]: Is it necessary to do extra logic if ender chest isnt tickable?

File: patches/server/0067-lithium-skip-ticking-block-entities-that-are-doing-n.patch [hunk 12]
```diff
@@ -0,0 +1,1348 @@
...
++                // Yatopia end
+                 this.world.playSound((EntityHuman) null, d0, (double) j + 0.5D, d2, SoundEffects.BLOCK_ENDER_CHEST_CLOSE, SoundCategory.BLOCKS, 0.5F, this.world.random.nextFloat() * 0.1F + 0.9F);
+             }, "Chest Sounds");
+ 
+@@ -79,11 +115,12 @@ public class TileEntityEnderChest extends TileEntity { // Paper - Remove ITickab
+                 this.a = 0.0F;
+             }
+         }
+-
++        this.checkWakeUp(); // Yatopia
```
Comment (ishland) [inline patches/server/0067-lithium-skip-ticking-block-entities-that-are-doing-n.patch]: Unnecessary diff

File: patches/server/0067-lithium-skip-ticking-block-entities-that-are-doing-n.patch [hunk 13]
```diff
@@ -0,0 +1,1348 @@
...
+index ab553f408b31cb7e2d281cc3d8fe42e19dc717d9..6e426d8e4c6e3768730b0d66267f44c1eee645ed 100644
+--- a/src/main/java/net/minecraft/server/TileEntitySkull.java
++++ b/src/main/java/net/minecraft/server/TileEntitySkull.java
+@@ -23,6 +23,7 @@ import com.mojang.authlib.Agent;
+ import com.mojang.authlib.ProfileLookupCallback;
+ import java.util.concurrent.Callable;
+ // Spigot end
++import me.jellysquid.mods.lithium.common.world.blockentity.BlockEntitySleepTracker;
+ 
+ public class TileEntitySkull extends TileEntity /*implements ITickable*/ { // Paper - remove tickable
```
Comment (ishland) [inline patches/server/0067-lithium-skip-ticking-block-entities-that-are-doing-n.patch]: Is it necessary to do extra logic if skull isnt tickable?

File: patches/server/0067-lithium-skip-ticking-block-entities-that-are-doing-n.patch [hunk 14]
```diff
@@ -0,0 +1,1348 @@
...
+ 
+@@ -150,7 +151,11 @@ public class TileEntitySkull extends TileEntity /*implements ITickable*/ { // Pa
+                 this.h = false;
+             }
+         }
+-
```
Comment (ishland) [inline patches/server/0067-lithium-skip-ticking-block-entities-that-are-doing-n.patch]: Unnecessary diff

File: patches/server/0067-lithium-skip-ticking-block-entities-that-are-doing-n.patch [hunk 15]
```diff
@@ -0,0 +1,1348 @@
...
+@@ -32,8 +32,11 @@ import org.bukkit.craftbukkit.block.CraftBlockState;
+ import org.bukkit.craftbukkit.block.data.CraftBlockData;
+ import org.bukkit.event.block.BlockPhysicsEvent;
+ // CraftBukkit end
++import me.jellysquid.mods.lithium.common.util.collections.MaskedTickingBlockEntityList;
++import me.jellysquid.mods.lithium.common.world.blockentity.BlockEntitySleepTracker;
++import me.jellysquid.mods.lithium.common.world.blockentity.SleepingBlockEntity;
+ 
+-public abstract class World implements GeneratorAccess, AutoCloseable {
++public abstract class World implements GeneratorAccess, AutoCloseable, BlockEntitySleepTracker {
```
Comment (ishland) [inline patches/server/0067-lithium-skip-ticking-block-entities-that-are-doing-n.patch]: Don't do mixin interface workaround here

**Review by ishland — CHANGES_REQUESTED** [review]
There are too many missing yatopia comments so I didn't point them all out.

**Comment by PaulBGD** [comment]
imo tile/block entities aren't actually slow, the only time they spend is on actual work (hoppers moving items from inventories for example.) I'm not sure it's worth taking on the maintenance of a patch that isn't proven to improve performance (in fact there's more overhead for all tile entities) for something that's not really an issue.

File: patches/server/0067-lithium-skip-ticking-block-entities-that-are-doing-n.patch (hunk 11 above)
Comment (Titaniumtown) [reply patches/server/0067-lithium-skip-ticking-block-entities-that-are-doing-n.patch]: Good catch.

File: patches/server/0067-lithium-skip-ticking-block-entities-that-are-doing-n.patch (hunk 13 above)
Comment (Titaniumtown) [reply patches/server/0067-lithium-skip-ticking-block-entities-that-are-doing-n.patch]: Nope

**Comment by Titaniumtown** [comment]
Closing: I am going to put this patch into another PR with more stuff I've ported.
